---
title: "Clustering & heatmap for Ace sequences in LNPA panle & frequencies of resistant Ace alleles in LNPA and GDRP panels "
author: "Marianna Karageorgi"
date: "2023-12-05"
---

## 1. Data

```{r}
rm(list=ls())

library(tidyverse)
library(ggdendro)
library(dendextend)
library(data.table)
library(grid)
```

-   The SNP table for the inbred lines used in Orchard 2021 can be found in the Google Drive in this link:

    <https://drive.google.com/drive/u/0/folders/1gw2rLUUoO_AT9DdcfINohVPSwA61vHLO>

    In the SNP dataframe, **columns** represent the inbred lines while **rows** represent the biallelic sites across the 3R chromosome.

-   Prepare a data frame with information on inbred line, sensitive/resistant Ace allele, and biallelic SNPs within the extended gene region of Ace. Convert the ATGC character representations of biallelic SNPs to numerical values, coded as 0 if they match the reference line, coded as 1 if alternative allele, coded as 0.5 if IUPAC letter for heterozygous.

-   The Ace coordinates of extended gene region on genome version R5:

3R:9,048,673..9,085,239 [-].

-   The coordinates of exons for genome version R5:

3R:complement(join(9050347..9055492,9058401..9058564,9058683..9058846,9062851..9063000,9063779..9063921,9068941..9069148,9069308..9069853,9069983..9070121,9071532..9072376,9084025..9084499))

-   The coordinates for the Ace resistance mutations for genome version R5:

chr3R: 9069721 (I161V), chr3R:9069408 (G265A), chr3R: 9069054 (F330Y), and chr3R:9063921 (G368A) using the reference genome v 5.39

### 1.1 Panel: S/R Ace alleles

```{r}
data = read.csv("./data/inbredv2_withHets.orch2021.3R.snpTable.npute")

# Filter rows based on the resistance mutation in column 'X3R' which corrensponds to positions
data.Ace.allele = data %>%
  filter(X3R %in% c(9069721, 9069408, 9069054, 9063921)) %>%
  
# Transpose the resulting data frame
  t() %>%
  
# Convert the transposed data frame to a regular data frame
  as.data.frame() %>%
  
# Set the column names based on the values from the first row
  setNames(unlist(.[1,])) %>%
  
# Convert the row names to a new column named "line"
  rownames_to_column(var = "line") %>%

# filter position line 
  filter(!line == "X3R") %>%
  
# Select only the columns 'Inbred_Line_ID', '9069721', '9069408', '9069054', and '9063921'
  select(line, `9069721`, `9069408`, `9069054`, `9063921`)

# Concatenate values from columns '9069721', '9069408', '9069054', and '9063921' into the new column 'Haplotype'
data.Ace.allele$haplotype = paste(data.Ace.allele$`9069721`, data.Ace.allele$`9069408`, data.Ace.allele$`9069054`, data.Ace.allele$`9063921`, sep = "") 

# Get unique haplotypes
unique_haplotypes = unique(data.Ace.allele$haplotype)

# Mutate 'Haplotype.aa' column based on conditions
data.Ace.haplo = data.Ace.allele %>%
   
  mutate(haplo = case_when(
    haplotype == "TCAC" ~ "IGFG.IGFG",
    haplotype == "CSWS" ~ "VGFA.VAYG",
    haplotype == "CCAS" ~ "VGFG.VGFA",
    haplotype == "TCWC" ~ "IGFG.IGYG",
    haplotype == "CGTC" ~ "VAYG.VAYG",
    haplotype == "CCAC" ~ "VGFG.VGFG",
    haplotype == "CCAG" ~ "VGFA.VGFA",
    haplotype == "YSWS" ~ "IGFG.VAYA",
    haplotype == "TCAS" ~ "IGFG.IGFA",
    haplotype == "YSAC" ~ "IGFG.VAFG",
    haplotype == "YSWC" ~ "IGFG.VAYG",
    haplotype == "YCAS" ~ "IGFG.VGFA",
    haplotype == "CSWC" ~ "VGFG.VAYG"
  )) %>%
  
  filter(!line == "Ref") %>%

  # Select lines & haplotypes
  select(line, haplo)


data.Ace.haplo.summary = data.Ace.haplo %>%
  group_by(haplo) %>%
  summarise(count = n())
```

#### 1.1.1 Table with allele frequencies LNPA

Table with sensitive and resistant Ace allele frequencies in the inbred lines (Extended Data Figure)

```{r}
library(kableExtra)
library(htmltools)
library(webshot)

# Count allele frequencies for figure 1/Extended Data Figure
allele.count.Ace = data.Ace.haplo %>%
  
  # Separate the Haplotype.aa column into two new columns (Allele.1 and Allele.2)
  separate(haplo, into = c("Allele.1", "Allele.2"), sep = "\\.") %>%
  
  # Pivot the data into a longer format to have a single column for haplotypes
  pivot_longer(cols = Allele.1:Allele.2, 
               names_to = "Allele", 
               values_to = "Ace_allele") %>%
  
  # Count occurrences of each haplotype
  dplyr::count(Ace_allele) %>%
  
  # Calculate frequency of each haplotype
  mutate(relative_frequency = round(n / sum(n) * 100, 2)) %>%
  
  # Keep only Ace allele & relative frequency, renaming columns for display
  select(`Ace allele` = Ace_allele, `Relative frequency (%)` = relative_frequency) 
  


# View the result
allele.count.Ace

#### Prepare table ####
table.allele.count.Ace = kable(x = allele.count.Ace, digits = 4, align = 'l',
                               # caption = "<b style='font-size:22px;'>Relative frequencies of sensitive and resistant <i>Ace</i> alleles in LNPA lines</b>",
                               escape = FALSE) %>%
                         kable_classic(full_width = TRUE, html_font = "Arial", font_size = 22) %>%
                         kable_styling(latex_options = c("striped", "hold_position"), position = "center") %>%
                         row_spec(0, bold = TRUE, font_size = 20) %>%  # Make header bold and set font size
                         row_spec(1:nrow(allele.count.Ace), font_size = 20)  # Smaller font for values

table.allele.count.Ace
```

#### 1.1.2 Table with allele frequencies LNPA & DGRP

```{r}
allele.count.Ace.dgrp = read_csv("./data/dgrp_Ace.csv") %>%
  rename('DGRP relative frequency (%)' = 'Relative frequency (%)')

allele.count.Ace.lnpa = allele.count.Ace %>%
  rename('LNPA relative frequency (%)' = 'Relative frequency (%)')

allele.count.Ace.all = full_join(allele.count.Ace.lnpa, allele.count.Ace.dgrp) %>%
    rename('Ace haplotype' = 'Ace allele') %>%
    arrange(desc(`LNPA relative frequency (%)`))

#### Prepare table ####
# Use kable to create the table
allele.count.Ace.all.kable = allele.count.Ace.all %>%
  kbl(caption = "") %>%
  kable_classic(full_width = TRUE, html_font = "Arial") 
  

allele.count.Ace.all.kable

# Save the table as an HTML file
save_kable(allele.count.Ace.all.kable, file = "./tables/table.allele.count.Ace.all.html")


# Use webshot to convert HTML to PDF
webshot("./tables/table.allele.count.Ace.all.html", file = "./tables/table.allele.count.Ace.all.pdf")
```

### 1.2 Panel: Biallelic SNPs in Ace region

```{r}
# Filter rows which corrensponds to positions within the Ace region
data.Ace.region = data %>%
  filter(between(X3R, 9048673, 9085239)) %>%
  
# Transpose the resulting data frame
  t() %>%
  
# Convert the transposed data frame to a regular data frame
  as.data.frame() %>%
  
# Set the column names based on the values from the first row
  setNames(unlist(.[1,])) %>%
  
# Convert the row names to a new column named "line"
  rownames_to_column(var = "line") %>%

# filter position line 
  filter(!line == "X3R") 
```

Convert the ATGC character representations of biallelic SNPs to numerical values, coded as 0 if they match the reference line, coded as 1 if alternative allele, coded as 0.5 if IUPAC letter for heterozygous.

```{r}
# The first row is the reference sequence

# Convert the SNP matrix to a character matrix if it's not already
data_matrix = as.matrix(data.Ace.region, stringsAsFactors = FALSE)

# Initialize a new matrix to hold the numeric values
numeric.data.Ace.region <- matrix(nrow = nrow(data_matrix), ncol = ncol(data_matrix))

# Assign the first column of the numeric matrix to be the same as the first column of the data matrix
numeric.data.Ace.region[, 1] = data_matrix[, 1]

# Define the IUPAC codes for heterozygous alleles
iupac_codes = c("R", "Y", "S", "W", "K", "M", "B", "D", "H", "V", "N")

# Iterate over each column (genomic position), starting from the second column
for (col in 2:ncol(data_matrix)) {
  ref_allele <- data_matrix[1, col] # Reference allele for this column
  
  # Iterate over each row (inbred line)
  for (row in 1:nrow(data_matrix)) {
    allele <- data_matrix[row, col]
    
    if (allele %in% iupac_codes) {
      # Heterozygous
      numeric.data.Ace.region[row, col] <- 0.5
    } else if (allele == ref_allele) {
      # Same as reference
      numeric.data.Ace.region[row, col] <- 0
    } else {
      # Alternative allele
      numeric.data.Ace.region[row, col] <- 1
    }
  }
}

# Assign row and column names to the new numeric matrix
colnames(numeric.data.Ace.region) = colnames(data_matrix)
rownames(numeric.data.Ace.region) = rownames(data_matrix)

# Turn numeric_data_matrix back to data.frame
numeric.data.Ace.region= as.data.frame(numeric.data.Ace.region)
```

### 1.3 Panel: S/R Ace allele & biallelic SNPs

```{r}
data.Ace = inner_join(data.Ace.haplo,numeric.data.Ace.region, by = "line") 
```

## 2. Heatmap

### 3.1 Heatmap Ace region

We will first change scale the resistance mutation alleles so as to be able to label with a different color. We scale them by 3. The coordinates for the Ace resistance mutations are chr3R: 9069721 (I161V), chr3R:9069408 (G265A), chr3R: 9069054 (F330Y), and chr3R:9063921 (G368A) using the reference genome v 5.39. So, alleles that are as the reference (0) will remain 0, alleles that are hetero (0.5) will turn to 1.5, and alleles that are the alternative (1) will turn to 3.

We will then transform the data to a "long" format, where each row only contains values for one measurement.

In addition, before the transformation,

```{r}
# Scale columns with the resistance mutations

# Define the columns you want to scale and multiply by 3
columns_to_scale = c("9069721", "9069408", "9069054", "9063921")

# Convert all columns except the first one to numeric
numeric.data.Ace.region <- numeric.data.Ace.region %>%
  mutate(across(-1, as.numeric))

# Apply the scaling and multiplication operation to the selected columns
numeric.data.Ace.region <- numeric.data.Ace.region %>%
  mutate(across(all_of(columns_to_scale), ~ . * 3))

# Heatmap
# Data wrangling
data_long = pivot_longer(data = numeric.data.Ace.region,
                              cols = -c(line),
                              names_to = "position",
                              values_to = "nucleotide")
```

Now we can use that new data frame, `data_long` to create our heatmap. In the `ggplot` package, we use the `geom_tile` layer for creating a heatmap. Given our prior experience with the y-axis labels being large, we will again use `theme` to make the accession numbers (the y-axis labels) a little smaller:

```{r}
# Convert "nucleotide" to a factor variable
data_long$nucleotide = as.factor(data_long$nucleotide)

# Create the heatmap plot
heatmap_plot = ggplot(data = data_long, aes(x = position, y = line, fill = nucleotide)) +
  geom_tile() +
  scale_fill_manual(values = c("grey50","gray70","grey90","yellow","orange"),
                    labels = c("ref", "het", "alt", "het_resist", "resist") 
                   # labels = c("0", "0.5", "1", "1.5", "3")
                    ) +
  scale_x_discrete("3R") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 90),
        axis.text.y = element_text(size = 6)) +
  ylab("Inbred Line ID")

heatmap_plot
```

### 3.2 Heatmap Ace exonic SNPs

Prepare a heatmap where we only keep the biallelic SNPs in the exons

```{r}
#Use mask to keep only exonic biallelic SNPs

## Define exon coordinates as start and end positions
exons = data.frame(start = c(9050347, 9058401, 9058683,9062851,9063779,9068941,9069308,9069983,9071532,9084025), end = c(9055492,9058564,9058846,9063000,9063921,9069148,9069853,9070121,9072376,9084499))

# Create a mask
mask = sapply(data_long$position, function(pos) {
  any(pos >= exons$start & pos <= exons$end)
})

# Filter exons using the mask
masked_data_long = data_long[mask, ] 
```

Now we can use that new data frame, `masked_data_long` to create our heatmap. In the `ggplot` package, we use the `geom_tile` layer for creating a heatmap. Given our prior experience with the y-axis labels being large, we will again use `theme` to make the accession numbers (the y-axis labels) a little smaller:

```{r}
# Convert "nucleotide" to a factor variable
masked_data_long$nucleotide = as.factor(masked_data_long$nucleotide)

# Create the heatmap plot
heatmap_plot2 = ggplot(data = masked_data_long, aes(x = position, y = line , fill = nucleotide)) +
  geom_tile() +
  scale_fill_manual(values = c("grey50","gray70","grey90","yellow","orange"),
                    labels = c("ref", "het", "alt", "het_resist", "resist") 
                   # labels = c("0", "0.5", "1", "1.5", "3")
                    ) +
  scale_x_discrete("3R") +
  ylab("Inbred Line ID") +

  theme(legend.position = "right",
        axis.text.x = element_text(size = 5, angle = 90),
        axis.text.y = element_text(size = 6)) 

heatmap_plot2
```

Re-order the heatmap with lines carrying S and R1, R2, R3 Ace alleles in homozygosity with corresponding color and then all heterozygous in grey

```{r}
# Define the order of the lines
line_order = c("LNPA.85", "LNPA.43_SP", "LNPA.36", "LNPA.29_SP", "X12LN6.4_B64", 
                         "X12LN6.41_B47", "X12LN10.89", "X12LN10.38", "X12LN10.32", 
                         "X12LN10.12", "X12LN10.10", "LNPA.37", "X12LN6.94_B19", 
                         "X12LN6.82_B4", "X12LN6.46_B37", "X12LN6.45_B26", "X12LN10.95", 
                         "X12LN10.43", "X12LN10.35", "LNPA.87", "LNPA.75_SP", 
                         "LNPA.40", "LNPA.17", "LNPA.15_12LN.15", "LNPA.14", 
                         "X12LN6.15_B41", "X12LN10.71_B", "LNPA.5_SP", "LNPA.53", 
                         "X12LN6.86_B29", "X12LN6.79_B81", "X12LN10.40", "LNPA.9", 
                         "LNPA.8_SP", "LNPA.7_SP", "LNPA.79", "LNPA.77", "LNPA.66", 
                         "LNPA.65_SP", "LNPA.62", "LNPA.61", "LNPA.58_SP", 
                         "LNPA.47_SP", "LNPA.44", "LNPA.38", "LNPA.34", 
                         "LNPA.31", "LNPA.3", "LNPA.2_SP", "LNPA.11", 
                         "X12LN6.92_B3", "X12LN6.70_B33", "X12LN6.66_B2", 
                         "X12LN6.50_B31", "X12LN6.49_B9", "X12LN6.31_B21", 
                         "X12LN6.30_B66", "X12LN6.2_B71", "X12LN6.28_B69", 
                         "X12LN6.16_B78", "X12LN6.12_B62", "X12LN10.97", 
                         "X12LN10.96", "X12LN10.86", "X12LN10.84", 
                         "X12LN10.83", "X12LN10.82", "X12LN10.80", 
                         "X12LN10.70", "X12LN10.67", "X12LN10.65", 
                         "X12LN10.59", "X12LN10.57", "X12LN10.29", 
                         "X12LN10.23", "X12LN10.13", "Ref")


# Order the lines using the line_order directly
masked_data_long$line = factor(x = masked_data_long$line, 
                               levels = line_order, 
                               ordered = TRUE)

# Convert "nucleotide" to a factor variable
masked_data_long$nucleotide = as.factor(masked_data_long$nucleotide)

# Create the heatmap plot
heatmap_plot2 = ggplot(data = masked_data_long, aes(x = position, y = line, fill = nucleotide)) +
  geom_tile() +
  scale_fill_manual(values = c("grey50","gray70","grey90","yellow","orange"),
                    labels = c("reference", "heterozygore", "alternative", "het_resistant", "resistant")) +
 
   
  scale_x_discrete("3R") +
  
  theme(
        axis.text.x = element_text(size = 4, angle = 90),
        axis.text.y = element_text(size = 4),
        axis.title.x = element_text(size = 12),
        axis.title.y = element_blank(),
        
        legend.position = "right",
        legend.text = element_text(size = 6),
        legend.title = element_blank(),
        legend.key.size = unit(0.4, "cm")) 
 

heatmap_plot2

ggsave("./plots/heatmap_LNPA.png", plot = heatmap_plot2, width = 8, height = 5)
```

## Appendix

### 1. How to use mask?

```{r}
# Example data
df <- data.frame(genomic_position = 1:100000)  # Replace with your actual data

# Define position frames as start and end positions
frames <- data.frame(start = c(1000, 5000, 10000), end = c(2000, 6000, 11000))

# Create a mask
mask <- sapply(df$genomic_position, function(pos) {
  any(pos >= frames$start & pos <= frames$end)
})

# Filter using the mask
filtered_df <- df[mask, ] 
```

### 2. Tutorial in dendrograms in R

<https://atrebas.github.io/post/2019-06-08-lightweight-dendrograms/>

How to color specific lines, see here:

<https://r-graph-gallery.com/31-custom-colors-in-dendrogram.html>

<http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning#change-labels>
